### PR TITLE
hifiasm module

### LINKED ISSUE #483: update module: hifiasm
I think it would be useful to have a module for this recently published genome assembler [hifiasm](https://bioconda.github.io/recipes/hifiasm/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [ ] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by sidorov-si**
## PR checklist

This module does not exist as yet in nf-core.

Closes #483 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/hifiasm/functions.nf [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/*
+ * -----------------------------------------------------
```
Comment (drpatelh) [inline modules/hifiasm/functions.nf:2]: Old version of `functions.nf`. Will need to re-install the latest version of tools.

File: modules/hifiasm/main.nf [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    if (use_parental_kmers) {
+        """
+        # Adding soft-links to original FastQs for consistent naming in a workflow
+        [ ! -f  ${prefix}.fastq.gz ] && ln -s ${reads} ${prefix}.fastq.gz
+
+        hifiasm $options.args \\
```
Comment (drpatelh) [inline modules/hifiasm/main.nf:48]:
```suggestion
        hifiasm \\
        $options.args \\
```

File: modules/hifiasm/main.nf [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            -o ${prefix}.asm \\
```
Comment (drpatelh) [inline modules/hifiasm/main.nf:49]:
```suggestion
        hifiasm \\
            $options.args \\
            -o ${prefix}.asm \\
```

File: modules/hifiasm/main.nf [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            ${prefix}.fastq.gz
+
+        hifiasm --version > ${software}.version.txt || exit 0
+        """
+    } else { // Phasing with Hi-C data is not supported yet
+        """
+        # Adding soft-links to original FastQs for consistent naming in a workflow
+        [ ! -f  ${prefix}.fastq.gz ] && ln -s ${reads} ${prefix}.fastq.gz
+
+        hifiasm $options.args \\
```
Comment (drpatelh) [inline modules/hifiasm/main.nf:62]:
```suggestion
        hifiasm \\
            $options.args \\
```

File: modules/hifiasm/main.nf [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            -o ${prefix}.asm \\
```
Comment (drpatelh) [inline modules/hifiasm/main.nf:63]:
```suggestion
        hifiasm \\
            $options.args \\
            -o ${prefix}.asm \\
```

File: modules/hifiasm/main.nf [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    tuple val(meta), path("*.r_utg.gfa")        , emit: raw_unitigs
+    tuple val(meta), path("*.p_utg.gfa")        , emit: processed_unitigs, optional: true
+    tuple val(meta), path("*.asm.p_ctg.gfa")    , emit: primary_contigs, optional: true
+    tuple val(meta), path("*.asm.a_ctg.gfa")    , emit: alternate_contigs, optional: true
+    tuple val(meta), path("*.hap1.p_ctg.gfa")   , emit: paternal_contigs, optional: true
+    tuple val(meta), path("*.hap2.p_ctg.gfa")   , emit: maternal_contigs, optional: true
+    tuple val(meta), path("*.ec.bin")           , emit: corrected_reads
+    tuple val(meta), path("*.ovlp.source.bin")  , emit: source_overlaps
+    tuple val(meta), path("*.ovlp.reverse.bin") , emit: reverse_overlaps
+    path  "*.version.txt"                       , emit: version
```
Comment (drpatelh) [inline modules/hifiasm/main.nf:37]:
```suggestion
    tuple val(meta), path("*.r_utg.gfa")       , emit: raw_unitigs
    tuple val(meta), path("*.ec.bin")          , emit: corrected_reads
    tuple val(meta), path("*.ovlp.source.bin") , emit: source_overlaps
    tuple val(meta), path("*.ovlp.reverse.bin"), emit: reverse_overlaps
    tuple val(meta), path("*.p_utg.gfa")       , emit: processed_unitigs, optional: true
    tuple val(meta), path("*.asm.p_ctg.gfa")   , emit: primary_contigs  , optional: true
    tuple val(meta), path("*.asm.a_ctg.gfa")   , emit: alternate_contigs, optional: true
    tuple val(meta), path("*.hap1.p_ctg.gfa")  , emit: paternal_contigs , optional: true
    tuple val(meta), path("*.hap2.p_ctg.gfa")  , emit: maternal_contigs , optional: true
    path  "*.version.txt"                      , emit: version
```

File: modules/hifiasm/main.nf [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    if (use_parental_kmers) {
+        """
+        # Adding soft-links to original FastQs for consistent naming in a workflow
+        [ ! -f  ${prefix}.fastq.gz ] && ln -s ${reads} ${prefix}.fastq.gz
```
Comment (drpatelh) [inline modules/hifiasm/main.nf:46]: Why do you need to rename here?

File: modules/hifiasm/main.nf [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+        hifiasm $options.args \\
+            -o ${prefix}.asm \\
+            -t $task.cpus \\
+            ${prefix}.fastq.gz
+
+        hifiasm --version > ${software}.version.txt || exit 0
```
Comment (drpatelh) [inline modules/hifiasm/main.nf:67]:
```suggestion
        echo $(hifiasm --version 2>&1) > ${software}.version.txt
```

File: modules/hifiasm/main.nf [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        [ ! -f  ${prefix}.fastq.gz ] && ln -s ${reads} ${prefix}.fastq.gz
+
+        hifiasm $options.args \\
+            -o ${prefix}.asm \\
+            -t $task.cpus \\
+            -1 $paternal_kmer_dump \\
+            -2 $maternal_kmer_dump \\
+            ${prefix}.fastq.gz
+
+        hifiasm --version > ${software}.version.txt || exit 0
```
Comment (drpatelh) [inline modules/hifiasm/main.nf:55]:
```suggestion
        echo $(hifiasm --version 2>&1) > ${software}.version.txt
```

File: tests/config/pytest_software.yml [hunk 10]
```diff
@@ -0,0 +1,820 @@
+abacas:
+  - software/abacas/**
+  - tests/software/abacas/**
```
Comment (drpatelh) [inline tests/config/pytest_software.yml:3]: This file has now been changed to `pytest_modules.yml` so you will need to use the latest version of tools or manually edit that file. This is why the tests aren't running.

File: tests/modules/hifiasm/main.nf [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+workflow test_hifiasm_hifi_only {
+
+    def input = []
+    input = [ [ id:'test' ], // meta map
+              [ file(params.test_data['homo_sapiens']['pacbio']['test_hifi_fastq_gz'], 
+              checkIfExists: true) ] 
+            ]
+    paternal_kmer_dump = file('dummy_paternal_yak')
+    maternal_kmer_dump = file('dummy_maternal_yak')
+    HIFIASM ( input, paternal_kmer_dump, maternal_kmer_dump, false )
```
Comment (drpatelh) [inline tests/modules/hifiasm/main.nf:19]:
```suggestion

    HIFIASM ( input, [], [], false )
```

**Review by drpatelh — COMMENTED** [review]
Looks great @sidorov-si. Some changes for consistency and others because it looks like you created the module using an older version of tools. May be easier to close this PR, re-create the module with the latest version of tools and then copy across the changes.

File: modules/hifiasm/functions.nf (hunk 1 above)
Comment (sidorov-si) [reply modules/hifiasm/functions.nf]: Yup, fixed!

File: modules/hifiasm/main.nf (hunk 7 above)
Comment (sidorov-si) [reply modules/hifiasm/main.nf]: You're right, I actually don't need this renaming :) Fixed!

File: tests/config/pytest_software.yml (hunk 10 above)
Comment (sidorov-si) [reply tests/config/pytest_software.yml]: Yup, fixed!

**Comment by sidorov-si** [comment]
Thank you very much for the review @drpatelh ! I believe I implemented all your comments.
The only thing that I left as it was is the version printing: your command with `$(..)` produces compilation error in Nextflow (not sure why).

**Comment by drpatelh** [comment]
Hi @sidorov-si ! I have copied across the final files in a new PR https://github.com/nf-core/modules/pull/596 because again it seems like you are pulling in commits that are merged already and hence the number of files changed is increasing whereas we should only have 6 files changed as you will see in that PR. With the way we are squashing commits not sure whether this will have an impact on things. I think it would be good for you to maybe get a handle on how to sync branches with `git`. Not entirely sure where you are going wrong but could ask others in the group how they are doing things? `git` is a bit of a beast!

**Pull request closed without merging**